contracts/contract_call.go

```diff
@@ -0,0 +1,120 @@
+package contracts
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/accounts/abi"
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/common/hexutil"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/log"
+)
+
+// Method represents a contract's method
+type Method struct {
+	abi    *abi.ABI
+	method string
+	maxGas uint64
+}
+
+// NewMethod creates a new contract message
+func NewMethod(abi *abi.ABI, method string, maxGas uint64) Method {
+	return Method{
+		abi:    abi,
+		method: method,
+		maxGas: maxGas,
+	}
+}
+
+func (am Method) Bind(contractAddress common.Address) *BoundMethod {
+	return &BoundMethod{
+		Method:         am,
+		resolveAddress: noopResolver(contractAddress),
+	}
+}
+
+// encodeCall will encodes the msg into []byte format for EVM consumption
+func (am Method) encodeCall(args ...interface{}) ([]byte, error) {
+	return am.abi.Pack(am.method, args...)
+}
+
+// decodeResult will decode the result of msg execution into the result parameter
+func (am Method) decodeResult(result interface{}, output []byte) error {
+	if result == nil {
+		return nil
+	}
+	return am.abi.Unpack(result, am.method, output)
+}
```

Comment:
These are private and only used once so I would suggest in-lining them

Reply:
i like them as they are :P

even if they add LOC, reading `encodeCall` and `decodeResult` tells a better story than just `am.abi.Unpack` `am.abi.Pack`.

Reply:
I'm not convinced about these functions telling a better story.

For me when I read the use of encodeCall I jumped to it's definition to see what it did, then I saw that it just calls abi.Pack and at that point I knew what it did, because I'm familiar with the go-ethereum codebase and I'd seen it before.

Having encodeCall adds one extra layer between the reader of the code and the underlying function Pack. Pack is much more likely to be known to anyone who is familar with the go-ethereum codebase and has more comprehensive documentation than encode call.

So I would say encodeCall actually obscures what is really happening from the reader. In my case I needed to jump to its definition to see what it did, whereas if we just called Pack directly I wouldn't have had to break my flow and figure out the meaning of encodeCall.

Reply:
Hi @mcortesi I will try to be more flexible in future. I see these conversations as allowing us to get to understand each other's coding values better, so its not just about the specific details here but also setting the scene for us to be able to come to conclusions on similar situations much more quickly in the future. So from that perspective too I don't consider this wasted time.